bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java

```diff
@@ -0,0 +1,926 @@
+            if (CTDConnected == true) {
+                initialize();
+                (new Thread(qEvents)).start();
+            } else {
+                handler.bridgeOffline("No communication with Qbus client");
+            }
+        } catch (IOException | InterruptedException e) {
+            logger.warn("Error initializing communication from thread {}", Thread.currentThread().getId());
+            // No connection with Qbus server, put Bridge offline
+            stopCommunication();
+            if (handler != null) {
+                handler.bridgeOffline("No communication with Qbus server");
+            }
+
+        }
+    }
+
+    /**
+     * Cleanup socket when the communication with Qbus Server is closed.
+     *
+     * @throws IOException
+     *
+     */
+    public synchronized void stopCommunication() {
+        listenerStopped = true;
+
+        Socket socket = qSocket;
+
+        if (socket != null) {
+            try {
+                socket.close();
+            } catch (IOException ignore) {
+                // ignore IO Error when trying to close the socket if the intention is to close it anyway
+            }
+        }
+
+        qSocket = null;
+
+        CTDConnected = false;
+        logger.debug("Communication stopped from thread {}", Thread.currentThread().getId());
+    }
+
+    /**
+     * Close and restart communication with Qbus Server.
+     */
+    public synchronized void restartCommunication() {
+        stopCommunication();
+
+        logger.debug("Qbus: restart communication from thread {}", Thread.currentThread().getId());
+
+        startCommunication();
+    }
+
+    /**
+     * Method to check if communication with Qbus Server is active
+     *
+     * @return True if active
+     */
+    public boolean communicationActive() {
+
+        return qSocket != null;
+    }
+
+    /**
+     * Method to check if communication with Qbus Client is active
+     *
+     * @return True if active
+     */
+
+    public boolean clientConnected() {
+        return CTDConnected;
+    }
+
+    /**
+     * Runnable that handles incomming communication from Qbus server.
+     * <p>
+     * The thread listens to the TCP socket opened at instantiation of the {@link QbusCommunication} class
+     * and interprets all incomming json messages. It triggers state updates for active channels linked to the
+     * Qbus outputs. It is started after initialization of the communication.
+     *
+     */
+    private Runnable qEvents = () -> {
+        String qMessage;
+
+        logger.info("Listening for events on thread {}", Thread.currentThread().getId());
+        listenerStopped = false;
+        qEventsRunning = true;
+
+        BufferedReader reader = qIn;
+
+        try {
+            if (reader == null) {
+                throw new IOException();
+            }
+            while (!listenerStopped & ((qMessage = reader.readLine()) != null)) {
```

Comment:
You could use the built-in interrupted flag. The thread is interrupted when you call `thread.interrupt()`.
```suggestion
            while (!Thread.currentThread().isInterrupted() & ((qMessage = reader.readLine()) != null)) {
```